File: Spring/Chapter4/java_demo/demo/src/main/java/com/jdk/JDKProxy.java
package com.jdk;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;

public class JDKProxy implements InvocationHandler {
    private JDKInterface testInterface;
    public Object createProxy(JDKInterface testInterface)
    {
        this.testInterface = testInterface;
        ClassLoader loader = JDKProxy.class.getClassLoader();
        Class<?>[] classes = testInterface.getClass().getInterfaces();
        return Proxy.newProxyInstance(loader,classes,this);
    }
    @Override
    public Object invoke(Object o, Method method, Object[] objects) throws Throwable {
        JDKAspect aspect = new JDKAspect();
        aspect.check();
        aspect.except();
        Object obj = method.invoke(testInterface,objects);
        aspect.log();
        aspect.monitor();
        return obj;
    }
}

File: Spring/Chapter4/kotlin_demo/demo/src/main/kotlin/jdk/JDKProxy.kt
package jdk

import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.lang.reflect.Proxy

class JDKProxy:InvocationHandler{
    lateinit var jdkInterface:JDKInterface

    fun createProxy(jdkInterface: JDKInterface): Any {
        this.jdkInterface = jdkInterface
        val loader:ClassLoader = JDKProxy::class.java.classLoader
        val classes:Array<Class<*>> = jdkInterface.javaClass.interfaces
        return Proxy.newProxyInstance(loader,classes,this)
    }

    override fun invoke(p0: Any?, p1: Method?, p2: Array<out Any>?): Any? {
        val aspect = JDKAspect()
        aspect.check()
        aspect.expect()
        val obj:Any? = p1?.invoke(jdkInterface,*(p2?: arrayOfNulls<Any>(0)))
        aspect.log()
        aspect.monitor()
        return obj
    }
}